src/AspNetCoreModuleV2/AspNetCore/applicationinfo.cpp

```diff
@@ -66,48 +68,76 @@ APPLICATION_INFO::CreateHandler(
 HRESULT
 APPLICATION_INFO::CreateApplication(const IHttpApplication& pHttpApplication)
 {
-    HRESULT hr = S_OK;
-
     if (AppOfflineApplication::ShouldBeStarted(pHttpApplication))
     {
         LOG_INFO(L"Detected app_offline file, creating polling application");
-        #pragma warning( push )
-        #pragma warning ( disable : 26409 ) // Disable "Avoid using new", using custom deleter here
+#pragma warning( push )
+#pragma warning ( disable : 26409 ) // Disable "Avoid using new", using custom deleter here
         m_pApplication.reset(new AppOfflineApplication(pHttpApplication));
-        #pragma warning( pop )
+#pragma warning( pop )
+        return S_OK;
     }
     else
     {
-        FINISHED_IF_FAILED(m_handlerResolver.GetApplicationFactory(pHttpApplication, m_pApplicationFactory));
+        try
+        {
+            const WebConfigConfigurationSource configurationSource(m_pServer.GetAdminManager(), pHttpApplication);
+            ShimOptions options(configurationSource);
 
-        LOG_INFO(L"Creating handler application");
-        IAPPLICATION * newApplication;
-        FINISHED_IF_FAILED(m_pApplicationFactory->Execute(
-            &m_pServer,
-            &pHttpApplication,
-            &newApplication));
+            auto hr = TryCreateApplication(pHttpApplication, options);
 
-        m_pApplication.reset(newApplication);
+            if (m_pApplication == nullptr || FAILED(hr))
+            {
+                // Log the failure and update application info to not try again
+                EventLog::Error(
+                    ASPNETCORE_EVENT_ADD_APPLICATION_ERROR,
+                    ASPNETCORE_EVENT_ADD_APPLICATION_ERROR_MSG,
+                    pHttpApplication.GetApplicationId(),
+                    hr);
+
+#pragma warning( push )
+#pragma warning ( disable : 26409 ) // Disable "Avoid using new", using custom deleter here
+                m_pApplication.reset(new ServerErrorApplication(
+                    pHttpApplication,
+                    hr,
+                    options.QueryHostingModel(),
+                    options.QueryDisableStartupPage()));
+#pragma warning( pop )
+            }
+        }
+        catch (ConfigurationLoadException &ex)
+        {
+            EventLog::Error(
+                ASPNETCORE_CONFIGURATION_LOAD_ERROR,
+                ASPNETCORE_CONFIGURATION_LOAD_ERROR_MSG,
+                ex.get_message().c_str());
+
+            m_pApplication.reset(new ServerErrorApplication(
+                pHttpApplication,
+                E_INVALIDARG /* hr */,
```

Comment:
this is pretty random

Reply:
Clarify?

Reply:
`E_INVALIDARG` why are you using this HR?

Reply:
E_FAIL instead?